src/sql/src/query_model/rewrite_engine.rs

```diff
@@ -0,0 +1,969 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::query_model::{
+    BoxId, BoxScalarExpr, BoxType, ColumnReference, DistinctOperation, Model, QuantifierId,
+    QuantifierSet, QuantifierType, Select,
+};
+use itertools::Itertools;
+use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
+
+/// Rule type
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum RuleType {
+    TopBoxOnly,
+    PreOrder,
+    PostOrder,
+}
+
+/// Trait for rewrite rules
+pub trait Rule {
+    fn name(&self) -> &'static str;
+
+    fn rule_type(&self) -> RuleType;
+
+    /// Whether the action should be fired for the given box.
+    /// This method is not allowed to modify the model in any way.
+    fn condition(&mut self, model: &Model, box_id: BoxId) -> bool;
+
+    /// Invoked immediately after `condition` if it returned true.
+    fn action(&mut self, model: &mut Model, box_id: BoxId);
+}
+
+/// Entry-point of the normalization stage.
+pub fn rewrite_model(model: &mut Model) {
+    let mut fixup_rules: Vec<Box<dyn Rule>> = vec![Box::new(Windowing::new())];
+    apply_rules_to_model(model, &mut fixup_rules);
+    model.garbage_collect();
+
+    let mut rules: Vec<Box<dyn Rule>> = vec![
+        Box::new(ColumnRemoval::new()),
+        Box::new(SelectMerge::new()),
+        Box::new(ConstantLifting::new()),
+        Box::new(WindowingToSelect::new()),
+    ];
+    apply_rules_to_model(model, &mut rules);
+    model.garbage_collect();
+
+    let mut decorrelation_rules: Vec<Box<dyn Rule>> = vec![Box::new(Decorrelation::new())];
+
+    apply_rules_to_model(model, &mut decorrelation_rules);
+    model.garbage_collect();
+
+    model.update_ids();
+}
+
+/// Transform the model by applying a list of rewrite rules.
+pub fn apply_rules_to_model(model: &mut Model, rules: &mut Vec<Box<dyn Rule>>) {
+    for rule in rules
+        .iter_mut()
+        .filter(|r| matches!(r.rule_type(), RuleType::TopBoxOnly))
+    {
+        apply_rule(&mut **rule, model, model.top_box);
+    }
+
+    deep_apply_rules(rules, model, model.top_box, &mut HashSet::new());
+
+    for rule in rules
+        .iter_mut()
+        .filter(|r| matches!(r.rule_type(), RuleType::TopBoxOnly))
+    {
+        apply_rule(&mut **rule, model, model.top_box);
+    }
+}
+
+/// Apply a rewrite rule to a given box within the Model if it matches the condition.
+fn apply_rule(rule: &mut dyn Rule, model: &mut Model, box_id: BoxId) {
```

Comment:
I have a question related to implementing the ScalarToForeach. This method seems to be about applying a rewrite rule to a box. But `ScalarToForeach` is about changing a quantifier. Which do you think is the best option?
1) There is a separate `apply_rule_to_quantifier` function that takes a QuantifierId instead of a BoxId.
2) `ScalarToForeach`  is applied to the input box of the quantifier to be changed.
3) `ScalarToForeach` is applied to the parent box of the quantifier to be changed.

Reply:
Definitely 3. A Quantifier always belongs in its parent box. So modifying it should be considered as a mutation of the parent box.

Reply:
Can you put your answer in a comment in the code?

Ah, since the answer is 3, so I would need to be remember which boxes satisfy the exactly-one-tuple condition so I know whether to change the quantifiers pointing to them when I get to the later boxes.